File: Java/DetectListCycle.java
/*
  Detect cycle in the list
  head pointer input could be NULL as well for empty list
  Node is defined as 
  class Node {
     int data;
     Node next;
  }
*/
int HasCycle(Node head) {
// mkberger
    Node tortoise = head;
    Node hare = head;
    int tortoise_position, hare_position;
    tortoise_position = hare_position = 0;
    if(head == null) return 0;
    while(true)
    {
        tortoise = tortoise.next;
        hare = hare.next;
        if (hare == null) return 0;
        hare = hare.next;
        if (hare == null) return 0;
        if(hare == tortoise) return 1;
    }
//
}

File: Linked-Lists/DetectListCycle.cpp
/*
  Detect loop in a linked list 
  List could be empty also
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/
int HasCycle(Node* head)
{
   // Complete this function
   // Do not write the main method
// mkberger
    Node* tortoise = new Node();
    Node* hare = new Node();
    tortoise = hare = head;
    if(head == nullptr) return 0;
    while(true)
    {
        tortoise = tortoise->next;
        hare = hare->next;
        if(hare == nullptr) return 0;
        hare = hare->next;
        if(hare == nullptr) return 0;
        if(tortoise == hare) return 1;
    }
//
}
